dask/dataframe/groupby.py

```diff
@@ -15,6 +15,112 @@
+# former do not shuffle the data and are efficiently implemented as tree
+# reductions. The latter is implemented by shuffling the underlying partiitons
+# such that all items of a group can be found in the same parititon.
+#
+# The argument to ``.groupby``, the index, can be a ``str``, ``dd.DataFrame``,
+# ``dd.Series``, or a list thereof. In operations on the grouped object, the
+# divisions of the the grouped object and the items of index have to align.
+# Currently, there is no support to shuffle the index values as part of the
+# groupby operation. Therefore, the alignment has to be guaranteed by the
+# caller.
+#
+# To operate on matchings paritions, most groupby operations exploit the
+# corresponding support in ``apply_concat_apply``. Specifically, this function
+# operates on matching paritiotns of frame-like objects passed as varargs. The
+# arguments to these functions are generated via ``__get_groupby_chunk_args``.
+#
+# After the inital chunk step, the passed index is implicitly passed along to
+# subsequent operations as the index of the parittions. Groupby operations on
+# the individual parttions can then access the index via the ``levels``
+# parameter of the ``groupby`` function. The correct arguments is determined by
+# the ``_determine_levels`` function.
+#
+# To minimize overhead, series in an index that were obtained by getitem on the
+# object to group are not passed as series to the various operations, but as
+# columnn keys. This transformation is implemented as ``_normalize_index``.
+#
+# Finally, the ``_get_index_meta`` allows to retrieve a index suitable for
+# usage in groupby operations of the meta objects. Both empty and nonempty meta
+# objects are supported.
+#
+# #############################################
+
+def _determine_levels(index):
+    """Determine the correct levels argument to groupby.
+    """
+    if isinstance(index, (tuple, list)) and len(index) > 1:
+        return list(range(len(index)))
+    else:
+        return 0
+
+
+def _normalize_index(df, index):
+    """Replace series with column names in an index wherever possible.
+    """
+    if not isinstance(df, DataFrame):
+        return index
+
+    elif isinstance(index, list):
+        return [_normalize_index(df, col) for col in index]
+
+    elif (isinstance(index, Series) and index.name in df.columns and
+          index._name == df[index.name]._name):
+            return index.name
+
+    elif (isinstance(index, DataFrame) and
+          set(index.columns).issubset(df.columns) and
+          index._name == df[index.columns]._name):
+        return list(index.columns)
+
+    else:
+        return index
+
+
+def _get_groupby_chunk_args(obj, index):
+    """Get the object to group and any partioned index item as list of varargs.
+    """
+    if not isinstance(index, list):
+        return [obj, index]
+    else:
+        return [obj] + index
+
+
+def _get_index_meta(index, meta_attribute):
```

Comment:
We might also want to inline this function.  It's slightly more complex but only used in two places.